File: old-app-java/src/main/java/tech/thdev/java_udemy_sample/data/source/image/ImageRepository.java
package tech.thdev.java_udemy_sample.data.source.image;

/**
 * Created by tae-hwan on 10/29/16.
 */

public class ImageRepository implements ImageSource {

    private ImageRemoteDataSource imageRemoteDataSource;

    private static ImageRepository imageSampleRepository;

    public static ImageRepository getInstance() {
        if (imageSampleRepository == null) {
            imageSampleRepository = new ImageRepository();
        }

        return imageSampleRepository;
    }

    private ImageRepository() {
        imageRemoteDataSource = ImageRemoteDataSource.getInstance();
    }

    @Override
    public void getImageItems(int page, final LoadImageCallback loadImageCallback) {
        imageRemoteDataSource.getImageItems(page, loadImageCallback);
    }
}

File: app/src/main/java/tech/thdev/app/data/source/image/ImageRepository.kt
package tech.thdev.app.data.source.image

import tech.thdev.app.data.ImageData

/**
 * Created by record-tae on 10/21/17.
 */
object ImageRepository : ImageDataSource {

    private val IMAGE_REMOTE_DATA: ImageRemoteData by lazy {
        ImageRemoteData()
    }

    override fun loadImageList(imageDataList: (List<ImageData>) -> Unit, size: Int) {
        IMAGE_REMOTE_DATA.loadImageList(imageDataList, size)
    }
}